### ISSUE TITLE
Consider an alternate http client?

### ISSUE THREAD

**Issue by wezm** [rhysd/git-brws#18]
I was just updating the [AUR package](https://aur.archlinux.org/packages/git-brws/) I maintain for `git-brws` and noticed that it seemed to have a lot of dependencies for a seemingly simple tool. I ran `cargo graph` on the project and it produced this:

![graph](https://user-images.githubusercontent.com/21787/68187589-25f1f980-fffb-11e9-911c-f89d665456b2.png)

The details aren't especially important but it shows that a large amount of dependencies come in via `reqwest` (top center of the graph). Now don't get me wrong, `reqwest` is great, but it is full featured, high performance, and asynchronous. Given `git-brws` only needs to make a very small number of requests per invocation it seems that it may be overkill.

Perhaps a simpler HTTP client would be a better choice. I did some searching on crates.io and found [minreq](https://docs.rs/minreq/2.0.0-alpha.2/minreq/), which seems like it might be a good fit for `git-brws` needs.

What do you think? If this is something you'd be open to I could look into making a PR.

**Comment by rhysd** [comment]
Thank you for suggesting. I thought the same after reading https://raphlinus.github.io/rust/2019/08/21/rust-bloat.html.

Yeah, this repository utilizes `reqwest` HTTPS client only for sending GET request. I chose `reqwest` for some reasons.

1. It was nicely implemented on solid fundamental libraries
2. Well tested
3. Async support
4. Well known and used widely

Regarding to 3., currently this repository only uses blocking request. But I think some requests can be sent asynchronously in the (near) future. Calling both `client.find_pr_url()` and `client.repo()` asynchronously would reduce latency when a forked repo is specified

https://github.com/rhysd/git-brws/blob/4cd0944948cc56061e3b91257d68ae8483e50be9/src/pull_request.rs#L48-L52

But it is a kind of 'all-in-one' package so it depends on so many packages as you kindly pointed. Replacing it would reduce build time and binary size (though they must be measured). It might be better to replace it for now until considering to send async request.

My current concerns are:

1. This repository supports HTTPS proxy. But minreq looks not supporting it
2. Quality. minreq seems not used widely. Number of tests are less

Other option is using libcurl with static link. Though it is kind of all-in-one library as you know, number of dependencies is small so build would be faster. My concern for this is introducing external C library (this tool supports Linux, macOS, Windows so we need to consider how to prepare libcurl on all platforms).

**Comment by rhysd** [comment]
FWIW, according to [changelog of reqwest](https://github.com/seanmonstar/reqwest/blob/master/CHANGELOG.md#v0100-alpha1), next major version will provide various features. I think we can also consider to omit features which we don't use.

**Comment by rhysd** [comment]
Reqwest 0.10.0 was released and I upgraded the dependency at b325d47.

Here is a summary:

* Before:
  - Dependencies: 186
  - Time to build: `cargo build  382.34s user 30.07s system 307% cpu 2:14.29 total`

* After:
  - Dependencies: 118
  - Time to build: `cargo build  301.18s user 20.98s system 305% cpu 1:45.46 total`

Build time was about 1.27x improved by omitting unnecessary features from reqwest crate. I also update this crate not to use blocking API so I did not opt-in blocking APIs of reqwest.

I measured build time for each crates with `cargo-bloat --time`:

```
  Time Crate
26.53s serde
24.00s syn
23.83s serde_derive
13.87s tokio
13.78s pin_project_internal
12.61s http
12.36s encoding_rs
12.04s futures_util
10.76s unicode_normalization
 8.61s reqwest
 8.12s hyper
 7.22s serde_json
 6.84s git_brws
 6.73s h2
 6.31s futures_macro
 6.05s rand
 5.80s mio
 5.77s mime_guess
 5.73s security_framework
 5.44s proc_macro2
 5.28s idna
 5.00s url
 4.44s proc_macro_hack
 4.35s c2_chacha
 4.18s ppv_lite86
 3.85s memchr
 3.27s libc
 3.27s httparse
 3.21s unicode_bidi
 3.18s tokio_macros
 3.03s getopts
 2.69s bytes
 2.49s autocfg
 2.22s time
 2.21s indexmap

...omitted
```

Though reqwest is still big crate, it is mainly for tokio. And tokio is necessary to manage async requests (for sending multiple requests at the same time).

It seems that now serde is a main part of bloat in this crate's dependencies. If we find a good alternative to serde for JSON parser, we can omit `json` feature of reqwest and can use the alternative.

With reqwest 0.10.0 improvements, I want to continue to use reqwest for now. I'm closing this issue but I may reopen this when I find better alternative to reqwest and tokio.
